### PR TITLE
Support named logos and omit logos by default

### PR DESCRIPTION
**Pull request by paulmelnikow**
- Except for social badges, omit logos by default (#983)
- Omit the logo from a social badge using the query string: `?logo=` (#983)
- Opt in to named logos using the query string: `?logo=appveyor`
- Provide custom logo data as before: `?logo=data:image/png;base64,...`
- Rewrite badge data functions, with unit tests

Unit tests are covering the new code very well, though the underlying functionality (logos) is untested.

### CODE REVIEW

**Comment by techtonik** [comment]
Many features in one go.

> Omit the logo from a social badge using the query string: ?logo= (#983)

I would merge this one immediately.

File: lib/badge-data.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  } else {
+    return [val];
+  }
+}
+
+function isDataUri(s) {
+  return s !== undefined && /^(data:)([^;]+);([^,]+),(.+)$/.test(s);
+}
+
+function hasPrefix (s, prefix) {
```
Comment (espadrine) [inline lib/badge-data.js]: For consistency's sake, could you avoid having a space between the function name and the opening parenthesis?
eg. `function hasPrefix(s, prefix)`.

**Comment by RedSparr0w** [comment]
> Could you look into why the Travis batch fails before I merge this?

Fairly sure this is just failing due to #979:
```
PR: badges/shields#1092
Error: 403 Forbidden
```

File: lib/badge-data.js (hunk 1 above)
Comment (paulmelnikow) [reply lib/badge-data.js]: Sure, will change them. I don't think debating is worth it. #1082 is a more productive way forward.

**Comment by paulmelnikow** [comment]
Thanks for the review! Have updated the PR. If you merge #1095 that will fix the test failure. 😄

**Comment by paulmelnikow** [comment]
Buh, I just closed this by accident. :( I'm going to go ahead and merge it into master. I've addressed the comments and the red build is due to #979.

**Comment by tooomm** [comment]
Shouldn't appveyor lose it's logo with this now?

Also, I can't find any information/details about that new behavior or named logo explanation etc. on the shields webpge.

**Comment by paulmelnikow** [comment]
Right, there should be some explanation on the homepage. I'll get a pull request open with that.

As far as appveyor, the logo is still there because this change hasn't been deployed yet.

**Comment by tylerl0706** [comment]
Nice! Just curious what logos are supported with these changes

**Comment by paulmelnikow** [comment]
Here's the list as of what is currently deployed: https://github.com/badges/shields/tree/gh-pages/logo

**Comment by RedSparr0w** [comment]
@techtonik think it is #1092

**Comment by techtonik** [comment]
@RedSparr0w it is #1092 already. =)

**Comment by tylerl0706** [comment]
@paulmelnikow Thanks for that! If I want to contribute an svg, are there any requirements? (single color, etc)

**Comment by paulmelnikow** [comment]
@tylerl0706 To stay on topic, let's discuss that in #1107.

**Comment by techtonik** [comment]
Requiring people to opt-in for logos by duplicating badge names looks crude - `https://img.shields.io/appveyor/ci/Eldinnie/python-telegram-bot/master.svg?logo=appveyor`. I still think the ideal way is to opt-out from them with `https://img.shields.io/appveyor/ci/Eldinnie/python-telegram-bot/master.svg?logo=`

**Comment by tooomm** [comment]
I understand that, for a clean and more uniform look of all badges, it defaults to off.
But an easier way to enable badges would be to allow `?logo=yes` or `?logo=on`.

The badge link basically contains its reference for a service logo already.
>img.shields.io/**appveyor**/ci

I'm pretty sure there is some pattern. :)

**Comment by paulmelnikow** [comment]
Nice idea. Or `?logo=default`. That would be a nice refinement, and not too hard to accomplish.

**Comment by techtonik** [comment]
I am still not convinced on defaults without logo - https://shields.io/ - all badges look the same without it.

**Pull request closed without merging**